service-base/src/main/java/org/eclipse/hono/service/tenant/BaseTenantService.java

```diff
@@ -0,0 +1,305 @@
+     *
+     * @param stopFuture the future to invoke once shutdown is complete.
+     */
+    public final void stop(final Future<Void> stopFuture) {
+        this.tenantConsumer.unregister();
+        this.log.info("unregistered tenant data consumer from event bus");
+        this.doStop(stopFuture);
+    }
+
+    /**
+     * Subclasses should override this method to perform any work required before shutting down this verticle.
+     * <p>
+     * This method is invoked by {@link #stop()} as part of the verticle deployment process.
+     * </p>
+     *
+     * @param stopFuture the future to invoke once shutdown is complete.
+     */
+    protected void doStop(final Future<Void> stopFuture) {
+        // to be overridden by subclasses
+        stopFuture.complete();
+    }
+
+    /**
+     * Processes a tenant request message received via the Vertx event bus.
+     *
+     * @param tenantMsg The message.
+     */
+    public final void processTenantMessage(final Message<JsonObject> tenantMsg) {
+        try {
+            final JsonObject body = tenantMsg.body();
+            final String tenantId = body.getString(RequestResponseApiConstants.FIELD_TENANT_ID);
+            final String operation = body.getString(MessageHelper.SYS_PROPERTY_SUBJECT);
+
+            if (tenantId == null) {
+                log.debug("tenant request does not contain mandatory property [{}]",
+                        RequestResponseApiConstants.FIELD_TENANT_ID);
+                reply(tenantMsg, TenantResult.from(HttpURLConnection.HTTP_BAD_REQUEST));
+                return;
+            }
+
+            JsonObject payload;
+
+            switch (operation) {
+            case ACTION_GET:
+                log.debug("retrieving tenant [{}]", tenantId);
+                get(tenantId, result -> reply(tenantMsg, result));
+                break;
+            case ACTION_ADD:
+                if (!isValidRequestPayload(body)) {
+                    log.debug("tenant request contains invalid structure");
+                    reply(tenantMsg, TenantResult.from(HttpURLConnection.HTTP_BAD_REQUEST));
+                    break;
+                }
+                payload = getRequestPayload(body);
+                log.debug("creating tenant [{}] with data {}", tenantId, payload != null ? payload.encode() : null);
+                add(tenantId, payload, result -> reply(tenantMsg, result));
+                break;
+            case ACTION_UPDATE:
+                payload = getRequestPayload(body);
+                log.debug("updating tenant [{}] with data {}", tenantId, payload != null ? payload.encode() : null);
+                update(tenantId, payload, result -> reply(tenantMsg, result));
+                break;
+            case ACTION_REMOVE:
+                log.debug("deleting tenant [{}]", tenantId);
+                remove(tenantId, result -> reply(tenantMsg, result));
+                break;
+            default:
+                break;
```

Comment:
the default should be to return a 400, shouldn't it?

Reply:
missed that :+1: